apps/src/applab/AppLabView.jsx

```diff
@@ -34,8 +35,101 @@ var AppLabView = React.createClass({
+    };
+  },
+
+  /**
+   * Called when the window resizes. Look to see if width/height changed, then
+   * call adjustHeights as our maxHeight may need adjusting.
+   */
+  onResize: function () {
+    var windowWidth = $(window).width();
+    var windowHeight = $(window).height();
+
+    // We fire window resize events when the grippy is dragged so that non-React
+    // controlled components are able to rerender the editor. If width/height
+    // didn't change, we don't need to do anything else here
+    if (windowWidth === this.state.width && windowHeight === this.state.height) {
+      return;
+    }
+
+    this.setState({
+      width: $(window).width(),
+      height: $(window).height()
+    });
+
+    this.adjustHeights();
+  },
+
+  /**
+   * Adjust the height/maxHeight of our instructions based on the rendered size
+   * of the instructions, and the rendered size of the workspace.
+   */
+  adjustHeights: function () {
+    if (!this.props.instructionsInTopPane) {
+      return;
+    }
+
+    var instructionsContent = this.refs.topInstructions.refs.instructions.refs.instructionsMarkdown;
+    var instructionsContentHeight = $(ReactDOM.findDOMNode(instructionsContent)).outerHeight(true);
+    var workspaceHeight = $(ReactDOM.findDOMNode(this.refs.codeWorkspace)).height();
+    var totalHeight = workspaceHeight + this.props.instructionsHeight;
+
+    var height = this.props.instructionsHeight || 300;
+    var maxHeight;
+
+    // Dont want height to extend past rendered length of content.
+    maxHeight = instructionsContentHeight + HEADER_HEIGHT + RESIZER_HEIGHT;
+    if (maxHeight < height) {
+      height = maxHeight;
+    }
+
+    // Have a preference for showing at least 150px of editor and 120px of
+    // debugger. Shrink instructions to make room. If that doesn't provide
+    // enough space, start also shrinking workspace
+    var EDITOR_RESERVE = 150;
+    var DEBUGGER_RESERVE = 120;
+    var INSTRUCTIONS_RESERVE = 150;
+
+    var topSpaceAvailable = totalHeight - EDITOR_RESERVE - DEBUGGER_RESERVE;
+    if (topSpaceAvailable < height) {
+      // if we'll still be at least 150px, just make our height smaller
+      if (topSpaceAvailable > INSTRUCTIONS_RESERVE) {
+        height = topSpaceAvailable;
+      } else {
+        height = Math.round(totalHeight / 3);
+      }
+      maxHeight = height;
+    } else if (topSpaceAvailable < maxHeight) {
+      maxHeight = topSpaceAvailable;
+    }
+    this.props.setInstructionsHeight(Math.min(height, maxHeight));
+    this.props.setInstructionsMaxHeight(maxHeight);
+  },
+
   componentDidMount: function () {
+    if (this.props.instructionsInTopPane) {
+      this.adjustHeights();
+
+      // Image loading can change the size of our instructions. Call adjustHeights
+      // so that our maxHeight is updated appropriately.
+      var dom = ReactDOM.findDOMNode(this.refs.topInstructions);
+      $(dom).find('img').load(function () {
+        this.adjustHeights();
+      }.bind(this));
```

Comment:
It might make sense to push this 'sensing image load events' work down into the `TopInstructions` component, and just pass it a `onContentDimensionsChanged` callback or something.